src/builtins.c

```diff
@@ -416,17 +419,29 @@ static uintptr_t NOINLINE jl_object_id__cold(jl_datatype_t *dt, jl_value_t *v) J
 
 JL_DLLEXPORT inline uintptr_t jl_object_id_(jl_value_t *tv, jl_value_t *v) JL_NOTSAFEPOINT
 {
+    uintptr_t *hash_value = (uintptr_t*)ptrhash_bp(&object_table, v);
+    if(*hash_value != (uintptr_t)HT_NOTFOUND) 
+        return *hash_value;
+    
     jl_datatype_t *dt = (jl_datatype_t*)tv;
-    if (dt == jl_symbol_type)
+    if (dt == jl_symbol_type) {
+        ptrhash_put(&object_table, v, (void *)((jl_sym_t*)v)->hash);
         return ((jl_sym_t*)v)->hash;
-    if (dt == jl_typename_type)
+    }
+    if (dt == jl_typename_type) {
+        ptrhash_put(&object_table, v, (void *)((jl_typename_t*)v)->hash);
         return ((jl_typename_t*)v)->hash;
+    }
     if (dt == jl_datatype_type) {
         jl_datatype_t *dtv = (jl_datatype_t*)v;
-        if (dtv->isconcretetype)
+        if (dtv->isconcretetype) {
+            ptrhash_put(&object_table, v, (void *)(dtv->hash));
             return dtv->hash;
+        }
     }
-    return jl_object_id__cold(dt, v);
+    *hash_value = jl_object_id__cold(dt, v);
```

Comment:
The call here to `jl_object_id__cold` makes the `hash_value` pointer invalid (you have the right idea on the next line, where you call `ptrhash_put`)